### PR TITLE
Upgrade rules

### PR DESCRIPTION
**Pull request by paulr34**

### CODE REVIEW

File: src-electron/db/query-session.js [hunk 1]
```diff
@@ -75,6 +75,27 @@ async function getSessionDirtyFlag(db, sessionId) {
...
+ * Resolves with true or false, depending whether this session is dirty.
+ *
+ * @export
+ * @param {*} db
+ * @param {*} sessionId
+ * @returns A promise that resolves into true or false, reflecting session dirty state.
+ */
+async function getSessionUpgradeFlag(db, sessionId) {
+  let row = await dbApi.dbGet(db, 'SELECT UPGRADE FROM UPGRADE')
+  if (row.UPGRADE == 0) {
```
Comment (paulr34) [inline src-electron/db/query-session.js:88]: I need to add another condition for null

File: src-electron/db/query-upgrade.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+/**
+ * Sets the upgrade status.
+ *
+ * @export
+ * @param {*} db
+ * @param {*} sessionId
+ * @returns A promise that resolves with the number of rows updated.
+ */
+async function setUpgradeStatus(db, status) {
```
Comment (paulr34) [inline src-electron/db/query-upgrade.js:74]: duplicate

File: src-electron/ide-integration/studio-rest-api.ts [hunk 3]
```diff
@@ -319,6 +323,28 @@ function sendDirtyFlagStatus(db: dbTypes.DbType) {
...
+    let status = await querySession.getSessionUpgradeStatus(
+      db,
+      session.sessionId
+    )
+    if (socket) {
+      if (upgrade == 1) {
+        wsServer.sendWebSocketMessage(socket, {
+          category: dbEnum.wsCategory.upgrade,
+          payload: status,
+        })
```
Comment (paulr34) [inline src-electron/ide-integration/studio-rest-api.ts:343]: when I merged a line was taken out which changes the UPGRADE entry to 0 - fixing now

File: src-electron/db/query-session.js [hunk 4]
```diff
@@ -75,6 +75,27 @@ async function getSessionDirtyFlag(db, sessionId) {
...
+    return 0
+  } else {
+    return 1
+  }
+}
+
+async function getSessionUpgradeStatus(db, sessionId) {
```
Comment (paulr34) [inline src-electron/db/query-session.js:95]: duplicate

File: src-electron/ide-integration/studio-rest-api.ts (hunk 3 above)
Comment (paulr34) [reply src-electron/ide-integration/studio-rest-api.ts:343]: oh, no. I forgot to update the ts file.

File: src/boot/ws.js [hunk 5]
```diff
@@ -128,13 +128,37 @@ onWebSocket(dbEnum.wsCategory.sessionCreationError, (data) => {
   </center>`
   Notify.create({
     message: html,
-    color: 'negative',
+    color: 'black',
```
Comment (jingteng25742) [inline src/boot/ws.js:131]: why are we changing the color?

File: src/boot/ws.js [hunk 6]
```diff
@@ -128,13 +128,37 @@ onWebSocket(dbEnum.wsCategory.sessionCreationError, (data) => {
...
     position: 'top',
     html: true,
     timeout: 0,
   })
 
-  console.log(`sessionCreationError: ${JSON.stringify(data)}`)
+  console.log(`upgrade ${JSON.stringify(data)}`)
```
Comment (jingteng25742) [inline src/boot/ws.js:137]: Debug / copy / paste error?

File: src-electron/db/db-mapping.js [hunk 7]
```diff
@@ -446,6 +446,13 @@ exports.map = {
       dirty: x.DIRTY == 1,
     }
   },
+  upgrade: (x) => {
+    if (x == null) return undefined
+    return {
+      upgrade: x.UPGRADE,
```
Comment (jingteng25742) [inline src-electron/db/db-mapping.js:507]: You can parse here via dbApi.fromDbBool(x.ATTRIBUTE_NAME), similar to other functions  and avoid parsing again in getSessionUpgradeFlag()

File: src-electron/db/query-session.js (hunk 1 above)
Comment (jingteng25742) [reply src-electron/db/query-session.js:88]: See above comment about parsing UPGRADE

File: src-electron/ide-integration/studio-rest-api.ts [hunk 8]
```diff
@@ -41,6 +41,7 @@ const op_add = '/rest/clic/component/add/project/'
 const op_remove = '/rest/clic/component/remove/project/'
 
 let dirtyFlagStatusId: NodeJS.Timeout
+let upgradeFlagStatusId: NodeJS.Timeout
```
Comment (jingteng25742) [inline src-electron/ide-integration/studio-rest-api.ts:46]: @paulr34 
Is this tested with Studio?

File: src/boot/ws.js [hunk 9]
```diff
@@ -128,13 +128,37 @@ onWebSocket(dbEnum.wsCategory.sessionCreationError, (data) => {
...
+//   // console.log(`Validation recieved: ${data}`)
+// })
+
+onWebSocket(dbEnum.wsCategory.upgrade, (data) => {
+  let html = `<center>
+  <strong>${data}</strong>
+  <br>
+  'upgrade'
+  </center>`
+  Notify.create({
```
Comment (jingteng25742) [inline src/boot/ws.js:152]: Let's add a Cypress unit test for coverage.

**Review by jingteng25742 — CHANGES_REQUESTED** [review]

File: src/boot/ws.js (hunk 9 above)
Comment (paulr34) [reply src/boot/ws.js:152]: Good idea, on it

File: src-electron/db/db-mapping.js (hunk 7 above)
Comment (paulr34) [reply src-electron/db/db-mapping.js:507]: okay thanks

File: src/boot/ws.js (hunk 6 above)
Comment (paulr34) [reply src/boot/ws.js:137]: yeah ill clean it up

File: src/boot/ws.js (hunk 5 above)
Comment (paulr34) [reply src/boot/ws.js:131]: no reason besides I thought it looked nice

File: src-electron/ide-integration/studio-rest-api.ts [hunk 10]
```diff
@@ -319,6 +325,21 @@ function sendDirtyFlagStatus(db: dbTypes.DbType) {
       })
     })
 }
+async function sendSessionUpgrade(db: dbTypes.DbType) {
+  let sessions = await querySession.getAllSessions(db)
+  for (const session of sessions) {
+    let socket = wsServer.clientSocket(session.sessionKey)
+    let rows = await queryNotice.getSessionNotices(db, session.sessionId)
+    if (socket) {
+      for (let i = 0; i < rows.length; i++) {
```
Comment (paulr34) [inline src-electron/ide-integration/studio-rest-api.ts:334]: I have not tested this out yet but this is the idea

**Comment by paulr34** [comment]
out of date

**Pull request closed without merging**